File: templates/problems/15_greedy/05_hand_of_straights/solution.go
package hand_of_straights

import "sort"

// solveIsNStraightHand uses sorting and a frequency map. O(n log n) time, O(n) space.
func solveIsNStraightHand(hand []int, groupSize int) bool {
	if len(hand)%groupSize != 0 {
		return false
	}
	freq := make(map[int]int)
	for _, card := range hand {
		freq[card]++
	}
	sort.Ints(hand)
	for _, card := range hand {
		if freq[card] == 0 {
			continue
		}
		for i := 0; i < groupSize; i++ {
			if freq[card+i] == 0 {
				return false
			}
			freq[card+i]--
		}
	}
	return true
}

File: templates/problems/15_greedy/05_hand_of_straights/typescript/solution.ts
/**
 * Use sorting and a frequency map. O(n log n) time, O(n) space.
 */
export function solveIsNStraightHand(hand: number[], groupSize: number): boolean {
    if (hand.length % groupSize !== 0) return false;
    const freq = new Map<number, number>();
    for (const card of hand) {
        freq.set(card, (freq.get(card) ?? 0) + 1);
    }
    hand = [...hand].sort((a, b) => a - b);
    for (const card of hand) {
        if ((freq.get(card) ?? 0) === 0) continue;
        for (let i = 0; i < groupSize; i++) {
            if ((freq.get(card + i) ?? 0) === 0) return false;
            freq.set(card + i, (freq.get(card + i) ?? 0) - 1);
        }
    }
    return true;
}
